common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java

```diff
@@ -0,0 +1,588 @@
+        }
+
+        Logger.verbose(
+                TAG + methodName,
+                "Found ["
+                        + allAccounts.size()
+                        + "] accounts."
+        );
+
+        return allAccounts;
+    }
+
+    @SuppressWarnings(UNCHECKED)
+    public AccountDeletionRecord removeAccountFromDevice(@NonNull final AccountRecord accountRecord) {
+        final Set<String> allClientIds = new HashSet<>();
+
+        allClientIds.addAll(mFociCache.getAllClientIds());
+        allClientIds.addAll(mAppUidCache.getAllClientIds());
+
+        for (final MsalOAuth2TokenCache optionalTokenCache : mOptionalCaches) {
+            allClientIds.addAll(optionalTokenCache.getAllClientIds());
+        }
+
+        final List<AccountDeletionRecord> deletionRecordList = new ArrayList<>();
+
+        for (final String clientId : allClientIds) {
+            deletionRecordList.add(
+                    removeAccount(
+                            accountRecord.getEnvironment(),
+                            clientId,
+                            accountRecord.getHomeAccountId(),
+                            null
+                    )
+            );
+        }
+
+        // Create a List of the deleted AccountRecords...
+        final List<AccountRecord> deletedAccountRecords = new ArrayList<>();
+
+        for (final AccountDeletionRecord accountDeletionRecord : deletionRecordList) {
+            deletedAccountRecords.addAll(accountDeletionRecord);
+        }
+
+        return new AccountDeletionRecord(deletedAccountRecords);
+    }
+
+    @Override
+    public AccountDeletionRecord removeAccount(String environment,
+                                               String clientId,
+                                               String homeAccountId,
+                                               @Nullable String realm) {
+        final String methodName = ":removeAccount";
+
+        AccountDeletionRecord deletionRecord = mAppUidCache.removeAccount(
+                environment,
+                clientId,
+                homeAccountId,
+                realm
+        );
+
+        Logger.verbose(
+                TAG + methodName,
+                "Accounts deleted count (uid): ["
+                        + deletionRecord.size()
+                        + "]"
+        );
+
+        if (deletionRecord.isEmpty()) {
+            deletionRecord = mFociCache.removeAccount(
+                    environment,
+                    clientId,
+                    homeAccountId,
+                    realm
+            );
+        }
+
+        Logger.verbose(
+                TAG + methodName,
+                "Accounts deleted count (foci): ["
+                        + deletionRecord.size()
+                        + "]"
+        );
+
+        final Iterator<MsalOAuth2TokenCache> cacheIterator = mOptionalCaches.iterator();
+
+        while (deletionRecord.isEmpty() && cacheIterator.hasNext()) {
```

Comment:
Can you help me understand the use case here too, deleting from other app's caches,

Reply:
This shouldn't have an effect (since the clientIds won't match) -- this should be removed in either case. Thanks for calling this out.

Reply:
There is also logic [here ](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java#L684-L694)which prevents an app from sticking a wildcard into this function. I'll add some javadoc to clarify on this.

Reply:
TODO: Add inline code comments to this function to document this behavior